x-pack/dockerlogbeat/magefile.go

```diff
@@ -47,132 +57,289 @@ func getPluginName() (string, error) {
+	}
+
+	mage.CreateDir(packageStagingDir)
+	mage.CreateDir(packageEndDir)
+	err = os.MkdirAll(filepath.Join(buildDir, "rootfs"), 0755)
+	if err != nil {
+		return errors.Wrap(err, "Error creating build dir")
+	}
 
-	err = sh.RunV("docker", "create", "--name", containerName, "rootfsimage", "true")
+	err = createContainer(cli)
+	if err != nil {
+		return errors.Wrap(err, "error creating base container")
+	}
+
+	// create the container that will become our rootfs
+	CreatedContainerBody, err := cli.ContainerCreate(context.Background(), &container.Config{Image: rootImageName}, nil, nil, "")
 	if err != nil {
 		return errors.Wrap(err, "error creating container")
 	}
 
-	err = sh.RunV("docker", "export", containerName, "-o", dockerExportPath)
+	defer func() {
+		// cleanup
+		if _, noClean := os.LookupEnv("DOCKERLOGBEAT_NO_CLEANUP"); !noClean {
+			err = cleanDockerArtifacts(CreatedContainerBody.ID, cli)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error cleaning up docker: %s", err)
+			}
+		}
+	}()
+
+	fmt.Printf("Got image: %#v\n", CreatedContainerBody.ID)
+
+	file, err := os.Create(dockerExportPath)
+	if err != nil {
+		return errors.Wrap(err, "error creating tar archive")
+	}
+
+	// export the container to a tar file
+	exportReader, err := cli.ContainerExport(context.Background(), CreatedContainerBody.ID)
 	if err != nil {
 		return errors.Wrap(err, "error exporting container")
 	}
 
-	err = mage.Copy("config.json", filepath.Join(packageStagingDir, "config.json"))
+	_, err = io.Copy(file, exportReader)
+	if err != nil {
+		return errors.Wrap(err, "Error writing exported container")
+	}
+
+	//misc prepare operations
+
+	err = mage.Copy("config.json", filepath.Join(buildDir, "config.json"))
 	if err != nil {
 		return errors.Wrap(err, "error copying config.json")
 	}
 
-	return sh.RunV("tar", "-xf", dockerExportPath, "-C", filepath.Join(packageStagingDir, "rootfs"))
+	// unpack the tar file into a root directory, which is the format needed for the docker plugin create tool
+	err = sh.RunV("tar", "-xf", dockerExportPath, "-C", filepath.Join(buildDir, "rootfs"))
+	if err != nil {
+		return errors.Wrap(err, "error unpacking exported container")
+	}
+
+	return nil
+}
+
+func cleanDockerArtifacts(containerID string, cli *client.Client) error {
+	fmt.Printf("Removing container %s\n", containerID)
+	err := cli.ContainerRemove(context.Background(), containerID, types.ContainerRemoveOptions{RemoveVolumes: true, Force: true})
+	if err != nil {
+		return errors.Wrap(err, "error removing container")
+	}
+
+	resp, err := cli.ImageRemove(context.Background(), rootImageName, types.ImageRemoveOptions{Force: true})
+	if err != nil {
+		return errors.Wrap(err, "error removing image")
+	}
+	fmt.Printf("Removed image: %#v\n", resp)
+	return nil
 }
 
-// CleanDocker removes working objects and containers
-func CleanDocker() error {
+// Uninstall removes working objects and containers
+func Uninstall() error {
```

Comment:
Also here, I think we could do:
```suggestion
func Uninstall(ctx context.Context) error {
```